File: Data-Structure/JS/Practice/sort-merge-sort.js
// merge
const merge = function (left, right) {

  let result = [];
  while(left.length > 0 && right.length > 0) {
    if(left[0] < right[0]) { 
      //if(left.shift() < right.shift()) {
      //do not use shift() compare, it remove element first
      result.push(left.shift());
    } else {
      result.push(right.shift());
    }
  }
  return [...result,...left,...right];
}
//console.log(merge([1,3,5,7],[2,4,6]));

// sort
const mergeSort = function (arr) {

  if (arr.length < 2) return arr;
  let mid = Math.floor(arr.length / 2);
  let left = mergeSort(arr.slice(0, mid));
  let right = mergeSort(arr.slice(mid));
  return merge(left, right);
}

console.log(mergeSort([0, 1, 5, 4, 3, 2, 6]));

File: Data-Structure/Python/sort/sort-merge-sort.py
def merge(left, right):

    if len(left) == 0:
        return right
    if len(right) == 0:
        return left
    arr = []

    while len(left) > 0 and len(right) > 0:
        if left[0] < right[0]:
            arr.append(left.pop(0))
        else:
            arr.append(right.pop(0))

    print('left=', left, 'right=', right, 'arr=', arr)
    arr.append(left)
    arr.append(right)
    return arr


def merge_sort(arr):
    if len(arr) < 2:
        return arr
    mid = len(arr) >> 1
    left = merge_sort(arr[: mid])
    right = merge_sort(arr[mid:])

    return merge(left, right)


print('merge_sort', merge_sort([1, 3, 5, 2, 4, 6]))
